File: scripts/lattice-cli/lattice_cli/preflight.py
"""
V2 preflight checks — runs BEFORE lattice deploy start to catch
environment issues early instead of failing mid-deployment.

Checks:
  [1] Docker daemon running and accessible without sudo
  [2] Docker Compose v2 plugin installed (not deprecated v1)
  [3] Host memory/CPU sufficient for Lattice OS workload
  [4] Required ports (3000, 5432, 6379) available
  [5] Disk space sufficient for image + volumes
  [6] Docker Hub auth (or skip for air-gapped)
  [7] License key valid and unexpired

Each check returns (passed, details) so the caller can decide
whether to continue or abort.
"""

import json
import os
import platform
import shutil
import socket
import subprocess
import sys
from typing import Any


def _run(cmd: list[str], timeout: int = 30) -> subprocess.CompletedProcess:
    try:
        return subprocess.run(
            cmd, capture_output=True, text=True, check=False, timeout=timeout
        )
    except FileNotFoundError:
        return subprocess.CompletedProcess(cmd, 127, "", f"command not found: {cmd[0]}")
    except subprocess.TimeoutExpired:
        return subprocess.CompletedProcess(cmd, 124, "", f"timeout after {timeout}s")


def check_docker_daemon() -> tuple[bool, str]:
    """[1] Is Docker daemon running and reachable without sudo?"""
    r = _run(["docker", "info"])
    if r.returncode == 0:
        # Parse version from output
        for line in r.stdout.splitlines():
            if "Server Version:" in line:
                ver = line.split(":", 1)[1].strip()
                return True, f"Docker {ver}"
        return True, "Docker running"
    if r.returncode == 127:
        return False, "Docker CLI not found"
    if "permission denied" in (r.stderr or "").lower():
        return False, "Docker requires sudo — add current user to docker group"
    if "cannot connect" in (r.stderr or "").lower():
        return False, "Docker daemon not running"
    return False, (r.stderr or r.stdout or "unknown error")[:200]


def check_compose_v2() -> tuple[bool, str]:
    """[2] Is Docker Compose v2 plugin installed (not deprecated v1)?"""
    # Check for deprecated docker-compose (v1)
    if shutil.which("docker-compose"):
        r = _run(["docker-compose", "--version"])
        if r.returncode == 0 and "v1" in (r.stdout + r.stderr).lower():
            return False, "Deprecated docker-compose v1 found — upgrade to Compose v2"

    r = _run(["docker", "compose", "version"])
    if r.returncode == 0:
        return True, r.stdout.strip()
    return False, "Docker Compose v2 not installed (https://docs.docker.com/compose/install/)"


def check_host_resources() -> tuple[bool, str]:
    """[3] Does host have sufficient memory/CPU?

    Lattice OS minimum requirements:
      - 4 GB RAM (8 GB recommended)
      - 2 CPU cores (4 recommended)
    """
    min_ram_gb = 4
    min_cpus = 2

    # Memory (cross-platform)
    ram_gb = 0.0
    if platform.system() == "Linux":
        try:
            with open("/proc/meminfo") as f:
                for line in f:
                    if line.startswith("MemTotal:"):
                        kb = int(line.split()[1])
                        ram_gb = kb / (1024 * 1024)
                        break
        except OSError:
            pass
    elif platform.system() == "Darwin":
        r = _run(["sysctl", "-n", "hw.memsize"])
        if r.returncode == 0:
            ram_gb = int(r.stdout.strip()) / (1024 ** 3)

    # CPU count
    cpu_count = os.cpu_count() or 1

    issues = []
    if ram_gb > 0 and ram_gb < min_ram_gb:
        issues.append(f"RAM {ram_gb:.1f}GB < {min_ram_gb}GB minimum")
    if cpu_count < min_cpus:
        issues.append(f"CPUs {cpu_count} < {min_cpus} minimum")

    if issues:
        return False, "; ".join(issues)

    summary = []
    if ram_gb > 0:
        summary.append(f"{ram_gb:.1f}GB RAM")
    summary.append(f"{cpu_count} CPUs")
    return True, ", ".join(summary) if summary else "OK"


def check_ports(ports: list[int] | None = None) -> tuple[bool, str]:
    """[4] Are required ports available (not in use by another service)?"""
    default_ports = [3000, 5432, 6379]  # app, supabase-postgres, redis
    target_ports = ports or default_ports
    blocked = []

    for port in target_ports:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.settimeout(1.0)
            if s.connect_ex(("127.0.0.1", port)) == 0:
                # Something is listening on this port — is it Lattice?
                # If Lattice is already running, that's not a block.
                blocked.append(port)

    if blocked:
        return False, f"port(s) {blocked} occupied by another service"
    return True, f"ports {target_ports} available"


def check_disk_space(min_gb: int = 8) -> tuple[bool, str]:
    """[5] Is there enough disk space for image pull + volumes?

    Lattice OS image + Supabase volumes typically require 8 GB.
    """
    try:
        usage = shutil.disk_usage("/")
        free_gb = usage.free / (1024 ** 3)
    except OSError:
        return True, "could not check disk space"

    if free_gb < min_gb:
        return False, f"{free_gb:.1f}GB free ({min_gb}GB required)"
    return True, f"{free_gb:.1f}GB free"


def check_docker_auth() -> tuple[bool, str]:
    """[6] Is Docker Hub authentication configured?

    Reads ~/.docker/config.json (managed by `docker login`).
    """
    config_path = os.path.expanduser("~/.docker/config.json")
    if not os.path.exists(config_path):
        return False, "not authenticated (run `lattice auth login`)"

    try:
        with open(config_path) as f:
            data = json.load(f)
    except (OSError, json.JSONDecodeError):
        return False, "could not read docker config"

    auths = data.get("auths", {})
    # Look for docker.io or index.docker.io or any registry
    known_registries = ["https://index.docker.io/v1/", "docker.io", "registry-1.docker.io"]
    for reg in known_registries:
        if reg in auths:
            return True, f"authenticated ({reg})"

    if auths:
        # They're authenticated to some registry (maybe a private one)
        first_reg = next(iter(auths)).split("/")[-1]
        return True, f"authenticated to {first_reg}"

    return False, "no auths found (run `lattice auth login`)"


def _port_name(port: int) -> str:
    return {3000: "app", 5432: "postgres", 6379: "redis"}.get(port, str(port))


def run_preflight(skip_auth: bool = False, skip_ports: bool = False,
                  custom_ports: list[int] | None = None) -> list[dict[str, Any]]:
    """Run all preflight checks. Returns a list of result dicts.

    Each dict: {"step": int, "name": str, "passed": bool, "detail": str, "fix": str}
    """
    checks = [
        (1, "Docker daemon",        check_docker_daemon,
            "Install Docker: https://docs.docker.com/get-docker/"),
        (2, "Docker Compose v2",    check_compose_v2,
            "Install Compose v2: https://docs.docker.com/compose/install/"),
        (3, "Host resources",       check_host_resources,
            "Increase RAM/CPU in Docker Desktop settings"),
        (4, "Port availability",    lambda: check_ports(custom_ports) if not skip_ports else (True, "skipped"),
            "Stop conflicting services or use --port to override"),
        (5, "Disk space",           check_disk_space,
            "Free up disk space: `docker system prune -af`"),
        (6, "Docker Hub auth",      lambda: check_docker_auth() if not skip_auth else (True, "skipped (air-gapped)"),
            "Run: lattice auth login"),
    ]

    results = []
    for step, name, fn, fix in checks:
        try:
            passed, detail = fn()
        except Exception as e:
            passed, detail = False, f"exception: {e}"
        results.append({
            "step": step,
            "name": name,
            "passed": passed,
            "detail": detail,
            "fix": fix,
        })

    return results


def print_preflight_report(results: list[dict[str, Any]]) -> bool:
    """Pretty-print preflight results. Returns True if all passed."""
    print(f"\n  Lattice OS Preflight")
    print(f"  {'─' * 50}")

    all_ok = True
    for r in results:
        icon = "✓" if r["passed"] else "✗"
        status = f"{icon} {r['detail']}"
        print(f"  [{r['step']}/6] {r['name']:<22} {status}")
        if not r["passed"]:
            all_ok = False
            print(f"         fix: {r['fix']}")

    print(f"\n  {'─' * 50}")
    if all_ok:
        print(f"  ✓ All preflight checks passed")
    else:
        failed = sum(1 for r in results if not r["passed"])
        print(f"  ✗ {failed} check(s) failed — resolve before deploying")
    print()

    return all_ok

File: packages/lattice-mcp-local/src/preflight.ts
/**
 * Host preflight checks (TypeScript port of Python preflight.py).
 *
 * All checks are read-only and safe to run from a coding agent context.
 */
import * as net from 'net';
import { execSync } from 'child_process';
import * as os from 'os';
import * as fs from 'fs';
import * as path from 'path';
import * as z from 'zod';

export const preflightResultSchema = z.object({
  step: z.number().int().positive(),
  name: z.string(),
  passed: z.boolean(),
  detail: z.string(),
  fix: z.string(),
});

export type PreflightResult = z.infer<typeof preflightResultSchema>;

function run(cmd: string[], timeout = 30): { code: number; stdout: string; stderr: string } {
  try {
    const out = execSync(cmd.join(' '), { encoding: 'utf8', timeout, stdio: ['pipe', 'pipe', 'pipe'] });
    return { code: 0, stdout: out, stderr: '' };
  } catch (e: any) {
    const code = e.status ?? 1;
    const stdout = e.stdout ?? '';
    const stderr = e.stderr ?? '';
    return { code, stdout, stderr };
  }
}

export function checkDockerDaemon(): PreflightResult {
  const r = run(['docker', 'info']);
  if (r.code === 0) {
    const ver = r.stdout.split('\n').find((l) => l.startsWith('Server Version:'));
    return { step: 1, name: 'Docker daemon', passed: true, detail: ver ? ver.split(':')[1].trim() : 'Docker running', fix: '' };
  }
  if (r.code === 127) return { step: 1, name: 'Docker daemon', passed: false, detail: 'Docker CLI not found', fix: 'Install Docker: https://docs.docker.com/get-docker/' };
  const stderr = (r.stderr || r.stdout || '').toLowerCase();
  if (stderr.includes('permission denied')) return { step: 1, name: 'Docker daemon', passed: false, detail: 'Docker requires sudo — add current user to docker group', fix: 'sudo usermod -aG docker $USER' };
  if (stderr.includes('cannot connect')) return { step: 1, name: 'Docker daemon', passed: false, detail: 'Docker daemon not running', fix: 'Start Docker Desktop or run: sudo systemctl start docker' };
  return { step: 1, name: 'Docker daemon', passed: false, detail: (r.stderr || r.stdout || 'unknown error').slice(0, 200), fix: 'Check Docker logs' };
}

export function checkComposeV2(): PreflightResult {
  if (fs.existsSync('/usr/local/bin/docker-compose') || fs.existsSync('/usr/bin/docker-compose')) {
    const r = run(['docker-compose', '--version']);
    if (r.code === 0 && /v1/i.test(r.stdout + r.stderr)) {
      return { step: 2, name: 'Docker Compose v2', passed: false, detail: 'Deprecated docker-compose v1 found', fix: 'Upgrade to Compose v2: https://docs.docker.com/compose/install/' };
    }
  }
  const r = run(['docker', 'compose', 'version']);
  if (r.code === 0) return { step: 2, name: 'Docker Compose v2', passed: true, detail: r.stdout.trim(), fix: '' };
  return { step: 2, name: 'Docker Compose v2', passed: false, detail: 'Docker Compose v2 not installed', fix: 'https://docs.docker.com/compose/install/' };
}

export function checkHostResources(): PreflightResult {
  const minRamGb = 4;
  const minCpus = 2;

  let ramGb = 0;
  if (process.platform === 'linux' && fs.existsSync('/proc/meminfo')) {
    const mem = fs.readFileSync('/proc/meminfo', 'utf8');
    const m = mem.match(/^MemTotal:\s+(\d+) kB$/m);
    if (m) ramGb = parseInt(m[1], 10) / (1024 * 1024);
  } else if (process.platform === 'darwin') {
    const r = run(['sysctl', '-n', 'hw.memsize']);
    if (r.code === 0) ramGb = parseInt(r.stdout.trim(), 10) / 1024 ** 3;
  }

  const cpus = os.cpus().length;
  const issues: string[] = [];
  if (ramGb > 0 && ramGb < minRamGb) issues.push(`RAM ${ramGb.toFixed(1)}GB < ${minRamGb}GB minimum`);
  if (cpus < minCpus) issues.push(`CPUs ${cpus} < ${minCpus} minimum`);

  if (issues.length) return { step: 3, name: 'Host resources', passed: false, detail: issues.join('; '), fix: 'Increase RAM/CPU in Docker Desktop settings' };
  const parts: string[] = [];
  if (ramGb > 0) parts.push(`${ramGb.toFixed(1)}GB RAM`);
  parts.push(`${cpus} CPUs`);
  return { step: 3, name: 'Host resources', passed: true, detail: parts.join(', ') || 'OK', fix: '' };
}

export function checkPorts(ports: number[] = [3000, 5432, 6379]): PreflightResult {
  const blocked: number[] = [];
  for (const port of ports) {
    try {
      const s = net.createConnection(port, '127.0.0.1');
      s.destroy();
      blocked.push(port);
    } catch {
      // port is open
    }
  }
  if (blocked.length) return { step: 4, name: 'Port availability', passed: false, detail: `port(s) ${blocked} occupied`, fix: 'Stop conflicting services or use --port to override' };
  return { step: 4, name: 'Port availability', passed: true, detail: `ports ${ports} available`, fix: '' };
}

export function checkDiskSpace(minGb = 8): PreflightResult {
  let freeGb = 0;
  try {
    const r = run(['df', '-g', '/']);
    const lines = r.stdout.split('\n');
    const last = lines[lines.length - 1];
    const cols = last.trim().split(/\s+/);
    freeGb = parseInt(cols[3], 10);
  } catch {
    return { step: 5, name: 'Disk space', passed: true, detail: 'could not check disk space', fix: '' };
  }
  if (freeGb < minGb) return { step: 5, name: 'Disk space', passed: false, detail: `${freeGb.toFixed(1)}GB free (${minGb}GB required)`, fix: 'Free up disk space: docker system prune -af' };
  return { step: 5, name: 'Disk space', passed: true, detail: `${freeGb.toFixed(1)}GB free`, fix: '' };
}

export function checkDockerAuth(): PreflightResult {
  const cfg = path.join(os.homedir(), '.docker', 'config.json');
  if (!fs.existsSync(cfg)) return { step: 6, name: 'Docker Hub auth', passed: false, detail: 'not authenticated', fix: 'Run: lattice auth login' };
  try {
    const data = JSON.parse(fs.readFileSync(cfg, 'utf8'));
    const auths = data.auths || {};
    const known = ['https://index.docker.io/v1/', 'docker.io', 'registry-1.docker.io'];
    for (const reg of known) if (auths[reg]) return { step: 6, name: 'Docker Hub auth', passed: true, detail: `authenticated (${reg})`, fix: '' };
    if (Object.keys(auths).length) {
      return { step: 6, name: 'Docker Hub auth', passed: true, detail: `authenticated to ${Object.keys(auths)[0].split('/').pop()}`, fix: '' };
    }
    return { step: 6, name: 'Docker Hub auth', passed: false, detail: 'no auths found', fix: 'Run: lattice auth login' };
  } catch {
    return { step: 6, name: 'Docker Hub auth', passed: false, detail: 'could not read docker config', fix: 'Run: lattice auth login' };
  }
}

export function runAllPreflight(skipAuth = false, skipPorts = false, customPorts?: number[]): PreflightResult[] {
  const checks: PreflightResult[] = [checkDockerDaemon(), checkComposeV2(), checkHostResources()];
  checks.push(skipPorts ? { step: 4, name: 'Port availability', passed: true, detail: 'skipped', fix: '' } : checkPorts(customPorts));
  checks.push(checkDiskSpace());
  if (skipAuth) checks.push({ step: 6, name: 'Docker Hub auth', passed: true, detail: 'skipped (air-gapped)', fix: '' });
  else checks.push(checkDockerAuth());
  return checks;
}
